File: tier2_intermediate/top_k_elements/solutions/p973_k_closest_points.py
"""P973: K Closest Points to Origin (Medium)

TODO: Implement solve() below.
Hint: Use a max-heap of size k keyed by negative distance squared.
"""
import sys
sys.path.insert(0, ".")
from src.utils import Problem, TestCase
import heapq


class Solution(Problem):
    name = "973. K Closest Points to Origin"
    test_cases = [
        TestCase(input=([[1,3],[-2,2]], 1), expected=[[-2,2]], label="example 1"),
        TestCase(input=([[3,3],[5,-1],[-2,4]], 2), expected=[[-2,4],[3,3]], label="example 2 (sorted)"),
    ]

    def solve(self, points: list[list[int]], k: int) -> list[list[int]]:
        heap: list[tuple[int, list[int]]] = []
        for p in points:
            dist = -(p[0] * p[0] + p[1] * p[1])
            heapq.heappush(heap, (dist, p))
            if len(heap) > k:
                heapq.heappop(heap)
        result = sorted([p for _, p in heap])
        return result


if __name__ == "__main__":
    Solution().run()

File: tier2_intermediate/top_k_elements/solutions/p973_k_closest_points.mjs
/**
 * P973: K Closest Points to Origin (Medium)
 * https://leetcode.com/problems/k-closest-points-to-origin/
 * Topics: Array, Math, Divide and Conquer, Geometry, Sorting, Heap (Priority Queue), Quickselect
 *
 * Given an array of points where points[i] = [xi, yi] represents a point on the X-Y plane and an integer k, return the k closest points to the origin (0, 0).
 * The distance between two points on the X-Y plane is the Euclidean distance (i.e., sqrt((x1 - x2)^2 + (y1 - y2)^2)).
 * You may return the answer in any order. The answer is guaranteed to be unique (except for the order that it is in).
 *
 * Example 1:
 *     Input: points = [[1,3],[-2,2]], k = 1
 *     Output: [[-2,2]]
 *     Explanation:
 *     The distance between (1, 3) and the origin is sqrt(10).
 *     The distance between (-2, 2) and the origin is sqrt(8).
 *     Since sqrt(8) < sqrt(10), (-2, 2) is closer to the origin.
 *     We only want the closest k = 1 points from the origin, so the answer is just [[-2,2]].
 *
 * Example 2:
 *     Input: points = [[3,3],[5,-1],[-2,4]], k = 2
 *     Output: [[3,3],[-2,4]]
 *     Explanation: The answer [[-2,4],[3,3]] would also be accepted.
 *
 * Constraints:
 *     - 1 <= k <= points.length <= 104
 *     - -104 <= xi, yi <= 104
 *
 * Hint: Use a max-heap of size k keyed by negative distance squared.
 */

import { readLine, readInts, readInt, writeInt, writeInts, writeString, writeBool } from '../../wasm_libs/js/io.mjs';

function solve(points, k) {
  return [...points]
    .sort((a, b) => a[0] * a[0] + a[1] * a[1] - (b[0] * b[0] + b[1] * b[1]))
    .slice(0, k)
    .sort((a, b) => a[0] - b[0] || a[1] - b[1]);
}

const flat = readInts();
const k = readInt();
const points = [];
for (let i = 0; i < flat.length; i += 2) points.push([flat[i], flat[i + 1]]);
const result = solve(points, k);
for (const p of result) {
  writeInts(p);
}
